[glean] Document the pings available out of the box [ci skip] (#1566)
mozilla-mobile/android-components: components/service/glean/docs/pings.md

# glean pings

> glean provides a set of pings that are assembled, out of the box, without any developer intervention.

Reviewer: Should we highlight again that this is only when collection is enabled?

Author: Good point! I've slightly rephrased this.

Every glean ping is in JSON format and contains a [common section](#the-common-information-section)
with shared information data, included under the `ping_info` key.
The following is a list of the pings, along with the conditions that triggers them.

- [`baseline` ping](baseline.md)
- `events` ping (TBD)
- `metrics` ping (TBD)

> ## The `common` information section

Reviewer 2: Should we just call this the `ping_info` section to avoid confusion?

> The following fields are always included in this section, for every ping.

Reviewer: ```suggestion
The following fields are always included in the `ping_info` section, for every ping.
```

> 

Reviewer: Do we want a section detailing when/how these are generated, cleared, and regenerated? For instance, I thought I remembered that there was a profile reset or opt-out mechanism that reset the clientId or reset  `seq` to 0 or something...

Author: Good point. We don't have, right now, anything to reset user-level metrics. They go away when user uninstalls the product using glean. I filed bug 1513616 in case some products want to have a way to clear stuff manually without uninstalling the app.

| Field name | Type | Description |
|---|---|---|

> | `ping_type` | String | The name of the ping |

Reviewer: ```suggestion
| `ping_type` | String | The name of the ping type (e.g. "baseline", "metrics") ||
```

| `app_build` | String | TBD |

> | `telemetry_sdk_build` | String | An identifier for the version of the glean library |

Reviewer: ```suggestion
| `telemetry_sdk_build` | String | The version of the glean library |
```

> | `client_id` | String |  A UUID identifying a profile and allowing user-oriented correlation of data |

Reviewer: Isn't `UUID` a top-level type we could use here?

> | `seq` | Counter | A running counter of the number of times this ping type has been sent |

Reviewer: ```suggestion
| `seq` | Counter | A running counter of the number of times pings of this type have been sent |
```

| `experiments` | Object | A dictionary of [active experiments](#the-experiments-object) |
| `start_time` | Datetime | The time of the start of collection of the data in the ping |

> | `end_time` | Datetime | The time of the end of collection of the data in the ping, and ping generation time (which is likely well before ping transmission time) |

Reviewer: ```suggestion
| `end_time` | Datetime | The time of the end of collection of the data in the ping. This is also the time this ping was generated and is likely well before ping transmission time |
```

| `profile_age` | Datetime | TBD |

### The `experiments` object
This object contains experiments, keyed by the experiment `id`. Each listed experiment contains the
`branch` the client is enrolled in and may contain a string to string map with additional data in the

> `extra` key. Both the `id` and `branch` are truncated to 30 characters.

Reviewer: Limits like these remind me that there might be implicit datatype limits that might be helpful to link to from this documentation. Maybe a section "These data types are explained in further detail in \[the data types documentation](...). Any limits or exemptions to limits are documented here with the fields that have or are exempt from the limitations."

Author: Great point. I just filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1513618) about having in-tree documentation for the different metric types.

# glean pings

> If data collection is enabled, glean provides a set of pings that are assembled, out of the box,

Reviewer 3: nit:  As a native English speaker the commas around `out of the box` are a little awkward here.  It should still be grammatically correct to say:

`If data collection is enabled, glean provides a set of pings that are assembled out of the box without any developer intervention.`

without any developer intervention.
Every glean ping is in JSON format and contains a [common section](#the-common-information-section)

> with shared information data, included under the `ping_info` key.

Reviewer 3: nit: comma after ` information data` not necessary.

> The following is a list of the pings, along with the conditions that triggers them.

Reviewer 3: nit:  comma after ` the pings` not necessary

- [`baseline` ping](baseline.md)
- `events` ping (TBD)
- `metrics` ping (TBD)

## The `ping_info` section
The following fields are always included in the `ping_info` section, for every ping.

| Field name | Type | Description |
|---|---|---|
| `ping_type` | String | The name of the ping type (e.g. "baseline", "metrics") |
| `app_build` | String | TBD |
| `telemetry_sdk_build` | String | The version of the glean library |
| `client_id` | UUID |  A UUID identifying a profile and allowing user-oriented correlation of data |
| `seq` | Counter | A running counter of the number of times pings of this type have been sent |
| `experiments` | Object | A dictionary of [active experiments](#the-experiments-object) |
| `start_time` | Datetime | The time of the start of collection of the data in the ping |
| `end_time` | Datetime | The time of the end of collection of the data in the ping. This is also the time this ping was generated and is likely well before ping transmission time |
| `profile_age` | Datetime | TBD |

All the metrics surviving application restarts (e.g. `client_id`, `seq`, ...) are removed once the
application using glean is uninstalled.

### The `experiments` object

> This object contains experiments, keyed by the experiment `id`. Each listed experiment contains the

Reviewer 3: nit: comma after ` experiments` not necessary.
